bam/reader.go

```diff
@@ -386,63 +340,82 @@ func parseAux(aux []byte) (aa []sam.Aux) {
 	return
 }
 
-// buildAux constructs a single byte slice that represents a slice of sam.Aux.
-func buildAux(aa []sam.Aux) (aux []byte) {
-	for _, a := range aa {
-		// TODO: validate each 'a'
-		aux = append(aux, []byte(a)...)
-		switch a.Type() {
-		case 'Z', 'H':
-			aux = append(aux, 0)
-		}
-	}
-	return
+// buffer is light-weight read buffer.
+type buffer struct {
+	off  int
+	data []byte
 }
 
-type errReader struct {
-	r   *bgzf.Reader
-	n   int
-	err error
+func (b *buffer) bytes(n int) []byte {
+	s := b.off
+	b.off += n
+	return b.data[s:b.off]
 }
 
-func (r *errReader) Read(p []byte) (int, error) {
-	if r.err != nil {
-		return 0, r.err
-	}
-	var n int
-	n, r.err = r.r.Read(p)
-	for n < len(p) && r.err == nil {
-		var _n int
-		_n, r.err = r.r.Read(p[n:])
-		n += _n
-	}
-	r.n += n
-	return n, r.err
+func (b *buffer) len() int {
+	return len(b.data) - b.off
+}
+
+func (b *buffer) discard(n int) {
+	b.off += n
 }
 
-type binaryReader struct {
-	r   *errReader
-	buf [4]byte
+func (b *buffer) readUint8() uint8 {
+	b.off++
+	return b.data[b.off-1]
 }
 
-func (r *binaryReader) readUint8() uint8 {
-	r.r.Read(r.buf[:1])
-	return r.buf[0]
+func (b *buffer) readUint16() uint16 {
+	return binary.LittleEndian.Uint16(b.bytes(2))
 }
 
-func (r *binaryReader) readUint16() uint16 {
-	r.r.Read(r.buf[:2])
-	return binary.LittleEndian.Uint16(r.buf[:2])
+func (b *buffer) readInt32() int32 {
+	return int32(binary.LittleEndian.Uint32(b.bytes(4)))
 }
 
-func (r *binaryReader) readInt32() int32 {
-	r.r.Read(r.buf[:4])
-	return int32(binary.LittleEndian.Uint32(r.buf[:4]))
+func (b *buffer) readUint32() uint32 {
+	return binary.LittleEndian.Uint32(b.bytes(4))
 }
 
-func (r *binaryReader) readUint32() uint32 {
-	r.r.Read(r.buf[:4])
-	return binary.LittleEndian.Uint32(r.buf[:4])
+// newBuffer returns a new buffer reading from the Reader's underlying bgzf.Reader and
+// updates the Reader's lastChunk field.
+func newBuffer(br *Reader) (*buffer, error) {
+	var buf [4]byte
+	n, err := io.ReadFull(br.r, buf[:4])
+	// br.r.Chunk() is only valid after the call the Read(), so this
+	// must come after the first read in the record.
+	tx := br.r.Begin()
+	if err != nil {
+		return nil, err
+	}
+	if n != 4 {
+		return nil, errors.New("bam: invalid record: short block size")
+	}
+	b := &buffer{data: buf[:4]}
+	size := int(b.readInt32())
+	b.off, b.data = 0, make([]byte, size)
+	n, err = io.ReadFull(br.r, b.data)
+	br.lastChunk = tx.End()
```

Comment:
For this to work as I was suggesting, it does need to go into a defer, but I've put you through enough pain. So I'll fix that.